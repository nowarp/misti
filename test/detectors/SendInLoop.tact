
contract Test {
    fun tWhile() { 
        let i: Int = 0;
        while (i < 10) {
            send(SendParameters{
                to: sender(),
                value: i,
            });
            i = i + 1;
        }
    }

    fun tRepeat() { 
        repeat (10) {
            send(SendParameters{
                to: sender(),
                value: 0,
            });
        }
    }

    fun tUntil() {
        let i: Int = 0;
        do {
            send(SendParameters{
                to: sender(),
                value: i,
            });
            i = i + 1;
        } until (i == 10);
    }

    fun tForeach() {
        let m: map<Int, Int> = emptyMap();
        foreach (k, v in m) {
            send(SendParameters{
                to: sender(),
                value: v,
            });
        }
    }

    fun tNested() {
        let i: Int = 0;
        while (i < 10) {
            repeat (10) {
                send(SendParameters{
                    to: sender(),
                    value: i,
                });
            }
            i = i + 1;
        }
    }
}

// Additional test

struct Msg {
    a: Int;
}


contract SendInLoop {
    owner: Address;
    a: Int;

    init(owner: Address) {
        self.owner = owner;
        self.a = 0;
    }

    //Send function inside a while loop

    fun exampleWhileLoop(limit: Int) {
        let i = 0;
        while (i < limit) {
            i += 1;
         let a =  send(SendParameters{
                to: self.owner,
                value: 0,
                bounce: false,
                body: Msg{ a: i }.toCell()
            });
        }
    }

        // self.reply inside a loop
    fun testReply() {
        let i: Int = 0;
        while (i < 5) {
            self.reply(Msg{ a: i }.toCell());
            i = i + 1;
        }
    }
}




