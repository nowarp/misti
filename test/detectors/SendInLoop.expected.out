[MEDIUM] SendInLoop: Send function called inside a loop
test/detectors/SendInLoop.tact:6:13:
  5 |         while (i < 10) {
> 6 |             send(SendParameters{
                  ^
  7 |                 to: sender(),
Help: Consider refactoring to avoid calling send functions inside loops
See: https://nowarp.io/tools/misti/docs/detectors/SendInLoop

[MEDIUM] SendInLoop: Send function called inside a loop
test/detectors/SendInLoop.tact:16:13:
  15 |         repeat (10) {
> 16 |             send(SendParameters{
                   ^
  17 |                 to: sender(),
Help: Consider refactoring to avoid calling send functions inside loops
See: https://nowarp.io/tools/misti/docs/detectors/SendInLoop

[MEDIUM] SendInLoop: Send function called inside a loop
test/detectors/SendInLoop.tact:26:13:
  25 |         do {
> 26 |             send(SendParameters{
                   ^
  27 |                 to: sender(),
Help: Consider refactoring to avoid calling send functions inside loops
See: https://nowarp.io/tools/misti/docs/detectors/SendInLoop

[MEDIUM] SendInLoop: Send function called inside a loop
test/detectors/SendInLoop.tact:37:13:
  36 |         foreach (k, v in m) {
> 37 |             send(SendParameters{
                   ^
  38 |                 to: sender(),
Help: Consider refactoring to avoid calling send functions inside loops
See: https://nowarp.io/tools/misti/docs/detectors/SendInLoop

[MEDIUM] SendInLoop: Send function called inside a loop
test/detectors/SendInLoop.tact:48:17:
  47 |             repeat (10) {
> 48 |                 send(SendParameters{
                       ^
  49 |                     to: sender(),
Help: Consider refactoring to avoid calling send functions inside loops
See: https://nowarp.io/tools/misti/docs/detectors/SendInLoop

[MEDIUM] SendInLoop: Send function called inside a loop
test/detectors/SendInLoop.tact:80:19:
  79 |             i += 1;
> 80 |          let a =  send(SendParameters{
                         ^
  81 |                 to: self.owner,
Help: Consider refactoring to avoid calling send functions inside loops
See: https://nowarp.io/tools/misti/docs/detectors/SendInLoop

[MEDIUM] SendInLoop: Send function called inside a loop
test/detectors/SendInLoop.tact:93:13:
  92 |         while (i < 5) {
> 93 |             self.reply(Msg{ a: i }.toCell());
                   ^
  94 |             i = i + 1;
Help: Consider refactoring to avoid calling send functions inside loops
See: https://nowarp.io/tools/misti/docs/detectors/SendInLoop